File: cpp/main/stage4/Stage4.cpp
/**
 * Make this code clean.
 */

#include "Stage4.hpp"
double Stage4::compute(std::vector<double> c, bool v, bool b)
{
  double t = 0;
  for (size_t i = 0; i < c.size(); ++i) {
    t += c[i];
  }
  int d = compute(c.size(), b, v);
  double d2 = d * t / 100;
  return t + d2;
}

int Stage4::compute(int n, bool b1, bool b2) {
  int p = 0;
  if (n >= 5 && n < 10) {
    p += 4;
  } else if (n >= 10) {
    p += 10;
  }
  if (b1) {
    // black friday discount
    p += 20;
  }

  if (b2) {
    // vip client
    p += 5;
  }
  return p;
}

File: typescript/src/stage4/Stage4.ts
//todo Make this code clean.
export class Stage4 {
  compute(c: number[], v: boolean, b: boolean): number {
    let t = 0;
    for (let i = 0; i < c.length; ++i) {
      t += c[i];
    }
    const d = this.computeDiscount(c.length, b, v);
    const d2 = (d * t) / 100;
    return t - d2;
  }

  computeDiscount(n: number, b1: boolean, b2: boolean): number {
    let p = 0;
    if (n >= 5 && n < 10) {
      p += 4;
    } else if (n >= 10) {
      p += 10;
    }
    if (b1) {
      // black friday discount
      p += 20;
    }

    if (b2) {
      // vip client
      p += 5;
    }
    return p;
  }
}
